File: Typecast/TCLocaTable.m
//
//  TCLocaTable.m
//  Typecast
//
//  Created by David Schweinsberg on 3/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCLocaTable.h"
#import "TCDirectoryEntry.h"
#import "TCDataInput.h"
#import "TCHeadTable.h"
#import "TCMaxpTable.h"

@implementation TCLocaTable

- (id)initWithDataInput:(TCDataInput *)dataInput
         directoryEntry:(TCDirectoryEntry *)entry
              headTable:(TCHeadTable *)head
              maxpTable:(TCMaxpTable *)maxp
{
    self = [super init];
    if (self)
    {
        self.directoryEntry = [entry copy];
        NSMutableArray *offsets = [[NSMutableArray alloc] initWithCapacity:[maxp numGlyphs] + 1];
        BOOL shortEntries = [head indexToLocFormat] == 0;
        if (shortEntries)
        {
            _factor = 2;
            for (int i = 0; i <= [maxp numGlyphs]; ++i)
                [offsets addObject:[NSNumber numberWithUnsignedShort:[dataInput readUnsignedShort]]];
        }
        else
        {
            _factor = 1;
            for (int i = 0; i <= [maxp numGlyphs]; ++i)
                [offsets addObject:[NSNumber numberWithInt:[dataInput readInt]]];
        }
        _offsets = offsets;
    }
    return self;
}

- (int32_t)offsetAtIndex:(NSUInteger)index
{
    return [_offsets[index] unsignedIntValue] * _factor;
}

- (NSString *)description
{
    NSMutableString *str = [NSMutableString stringWithFormat:
                            @"'loca' Table - Index To Location Table\n--------------------------------------\n"
                            @"Size = %d bytes, %ld entries\n",
                            [[self directoryEntry] length],
                            [_offsets count]];
    NSUInteger i = 0;
    for (NSNumber *offset in _offsets)
    {
        [str appendFormat:@"        Idx %ld  -> glyfOff 0x%x\n", i, [self offsetAtIndex:i]];
        ++i;
    }
    return str;
}

@end

File: Typecast/TCLocaTable.swift
//
//  TCLocaTable.swift
//  Type Designer
//
//  Created by David Schweinsberg on 8/6/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation
import IOUtils

class TCLocaTable: TCTable, Codable {
  let offsets: [Int]
  let factor: Int
  let dataCount: Int

  init(data: Data, headTable: TCHeadTable, maxpTable: TCMaxpTable) {
    dataCount = data.count
    let dataInput = TCDataInput(data: data)
    var offsets = [Int]()
    let shortEntries = headTable.indexToLocFormat == 0
    if shortEntries {
      factor = 2
      for _ in 0...maxpTable.numGlyphs {
        offsets.append(Int(dataInput.readUInt16()))
      }
    } else {
      factor = 1
      for _ in 0...maxpTable.numGlyphs {
        offsets.append(Int(dataInput.readUInt32()))
      }
    }
    self.offsets = offsets;
    super.init()
  }

  func offset(at index: Int) -> Int {
    return offsets[index] * factor
  }

  override class var tag: TCTable.Tag {
    get {
      return .loca
    }
  }

  override var description: String {
    get {
      var str = String(format:
        "'loca' Table - Index To Location Table\n--------------------------------------\n" +
        "Size = %d bytes, %ld entries\n",
        dataCount,
        offsets.count)
      var i = 0
      for _ in offsets {
        str.append(String(format: "        Idx %ld  -> glyfOff 0x%x\n", i, offset(at: i)))
        i += 1
      }
      return str;
    }
  }
}
